src/Custom/Microsoft.Extensions.AI/OpenAIChatClient.cs

```diff
@@ -0,0 +1,760 @@
+
+            foreach (var entry in functionCallInfos)
+            {
+                FunctionCallInfo fci = entry.Value;
+                if (!string.IsNullOrWhiteSpace(fci.Name))
+                {
+                    var callContent = ParseCallContentFromJsonString(
+                        fci.Arguments?.ToString() ?? string.Empty,
+                        fci.CallId!,
+                        fci.Name!);
+                    responseUpdate.Contents.Add(callContent);
+                }
+            }
+
+            // Refusals are about the model not following the schema for tool calls. As such, if we have any refusal,
+            // add it to this function calling item.
+            if (refusal is not null)
+            {
+                (responseUpdate.AdditionalProperties ??= [])[nameof(ChatMessageContentPart.Refusal)] = refusal.ToString();
+            }
+
+            // Propagate additional relevant metadata.
+            if (fingerprint is not null)
+            {
+                (responseUpdate.AdditionalProperties ??= [])[nameof(ChatCompletion.SystemFingerprint)] = fingerprint;
+            }
+
+            yield return responseUpdate;
+        }
+    }
+
+    private static ChatResponse FromOpenAIChatCompletion(ChatCompletion openAICompletion, ChatOptions? options, ChatCompletionOptions chatCompletionOptions)
+    {
+        Argument.AssertNotNull(openAICompletion, nameof(openAICompletion));
+
+        // Create the return message.
+        ChatMessage returnMessage = new()
+        {
+            MessageId = openAICompletion.Id, // There's no per-message ID, so we use the same value as the response ID
+            RawRepresentation = openAICompletion,
+            Role = FromOpenAIChatRole(openAICompletion.Role),
+        };
+
+        // Populate its content from those in the OpenAI response content.
+        foreach (ChatMessageContentPart contentPart in openAICompletion.Content)
+        {
+            if (ToAIContent(contentPart) is AIContent aiContent)
+            {
+                returnMessage.Contents.Add(aiContent);
+            }
+        }
+
+        // Output audio is handled separately from message content parts.
+        if (openAICompletion.OutputAudio is ChatOutputAudio audio)
+        {
+            string mimeType = chatCompletionOptions?.AudioOptions?.OutputAudioFormat.ToString()?.ToLowerInvariant() switch
+            {
+                "opus" => "audio/opus",
+                "aac" => "audio/aac",
+                "flac" => "audio/flac",
+                "wav" => "audio/wav",
+                "pcm" => "audio/pcm",
+                "mp3" or _ => "audio/mpeg",
+            };
+
+            var dc = new DataContent(audio.AudioBytes.ToMemory(), mimeType)
+            {
+                AdditionalProperties = new() { [nameof(audio.ExpiresAt)] = audio.ExpiresAt },
+            };
+
+            if (audio.Id is string id)
+            {
+                dc.AdditionalProperties[nameof(audio.Id)] = id;
```

Comment:
I thought the plan was to avoid setting `AdditionalProperties` for data that could also be reached in a strongly-typed way through `RawRepresentation` (except if there's some reason to believe it needs to show up in telemetry, which I don't think is the case for the data here, and especially not for `Transcript` which might be huge).